# Introduction to Data Science 2025

# Week 1

## Exercise 1 | Matrix warm-up
<span style="background-color: #ccfff2"> *Note: You can find tutorials for NumPy and Pandas under 'Useful tutorials' in the course material.*</span>

One of the most useful properties of any scientific programming language (Python with NumPy, R, Julia, etc) is that they allow us to work with matrices efficiently. Let's learn more about these features!

### 1.1 Basics

1. Let's start by creating two arrays <span style="background-color: #ccfff2"> A</span> and <span style="background-color: #ccfff2"> B</span> which each have the integers <span style="background-color: #ccfff2"> 0, 1, 2, ..., 1e7-1</span>. Use the normal arrays or lists of the programming language you are using, e.g. *list* or *[ ]* or *numpy.array()* in Python.

In [1]:
A = list(range(0,int(1e7)))
B = list(range(0,int(1e7)))

In [2]:
A

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


2. Create a function that uses a <span style="background-color: #ccfff2"> for loop</span> or equivalent to return a new array <span style="background-color: #ccfff2"> C</span>, which contains the <span style="background-color: #ccfff2"> element-wise sum of *A* and *B*</span>, e.g. C should contain the integers <span style="background-color: #ccfff2"> 0, 2, 4, etc</span>.

In [3]:
def add_with_for(A,B):
    C = []
    # add your code here
    for i in range(len(A)):
        C.append(A[i]+B[i])
    return C

In [4]:
out = add_with_for(A,B)


In [5]:
out

[0,
 2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100,
 102,
 104,
 106,
 108,
 110,
 112,
 114,
 116,
 118,
 120,
 122,
 124,
 126,
 128,
 130,
 132,
 134,
 136,
 138,
 140,
 142,
 144,
 146,
 148,
 150,
 152,
 154,
 156,
 158,
 160,
 162,
 164,
 166,
 168,
 170,
 172,
 174,
 176,
 178,
 180,
 182,
 184,
 186,
 188,
 190,
 192,
 194,
 196,
 198,
 200,
 202,
 204,
 206,
 208,
 210,
 212,
 214,
 216,
 218,
 220,
 222,
 224,
 226,
 228,
 230,
 232,
 234,
 236,
 238,
 240,
 242,
 244,
 246,
 248,
 250,
 252,
 254,
 256,
 258,
 260,
 262,
 264,
 266,
 268,
 270,
 272,
 274,
 276,
 278,
 280,
 282,
 284,
 286,
 288,
 290,
 292,
 294,
 296,
 298,
 300,
 302,
 304,
 306,
 308,
 310,
 312,
 314,
 316,
 318,
 320,
 322,
 324,
 326,
 328,
 330,
 332,
 334,
 336,
 338,
 340,
 342,
 344,
 346,
 348,
 350,

3. Next, let's create another function that uses NumPy (or equivalent) to do the same. To try it out, allocate two arrays (e.g. using <span style="background-color: #ccfff2"> np.array</span> in NumPy) and add the arrays together using your function. Don't use loops, instead, find out how to add the two arrays directly. What do you notice in comparison to the previous function?

In [6]:
# Use this cell for your code
import numpy as np
A = np.array(range(0,int(1e7)))
B = np.array(range(0,int(1e7)))

C = A+B

In [7]:
C

array([       0,        2,        4, ..., 19999994, 19999996, 19999998])

### Its much fasted than loops. Also we do not need to think about the indexing.

### 1.2 Array manipulation

<span style="background-color: #ccfff2"> *Note: for the following exercises, only use NumPy or equivalent functions. Don't use any loops.* </span>
1. Create the following array:

*[hint: <span style="background-color: #ccfff2"> np.reshape</span>]*

In [8]:
# array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
#        [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
#        [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
#        [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
#        [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
#        [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
#        [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
#        [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
#        [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
#        [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

In [9]:
# Use this cell for your code
A = np.arange(0,100)
A2 = np.reshape(A,(10,10))
A2

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
       [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
       [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
       [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
       [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
       [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

2. Create the following array:

In [10]:
# array([[0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.]])

In [11]:
# Use this cell for your code
A2[A2 %2==1]=1
A2[A2 %2==0]=0

In [12]:
A2

array([[0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0, 1, 0, 1]])

3. Create the following array (D):

In [13]:
# array([[0., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 1., 0.]])

In [14]:
# Use this cell for your code
D = np.ones((10,10), dtype=int)
#A2 = np.reshape(A2,(10,10))
np.fill_diagonal(D, 0)
D

array([[0, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 0, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 0, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 0, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 0, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 0, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 0, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 0, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 0, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 0]])

4. Create the following array (E):

In [15]:
# array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [0., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

In [16]:
# Use this cell for your code
E = np.ones((10,10), dtype=int)
#A2 = np.reshape(A2,(10,10))
np.fill_diagonal(np.fliplr(E), 0)
E

array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [1, 1, 1, 1, 1, 1, 1, 1, 0, 1],
       [1, 1, 1, 1, 1, 1, 1, 0, 1, 1],
       [1, 1, 1, 1, 1, 1, 0, 1, 1, 1],
       [1, 1, 1, 1, 1, 0, 1, 1, 1, 1],
       [1, 1, 1, 1, 0, 1, 1, 1, 1, 1],
       [1, 1, 1, 0, 1, 1, 1, 1, 1, 1],
       [1, 1, 0, 1, 1, 1, 1, 1, 1, 1],
       [1, 0, 1, 1, 1, 1, 1, 1, 1, 1],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 1]])

5. Call the last two matrices <span style="background-color: #ccfff2">D</span> and <span style="background-color: #ccfff2">E</span>, respectively. Show that the determinant of their product (matrix multiplication) is the same as the product of their determinants. That is calculate both <span style="background-color: #ccfff2">det(DE)</span> and <span style="background-color: #ccfff2">det(D) * det(E)</span>, and show that they are the same. Is it a coincidence? (I think not) The product of the determinants (or the determinant of the product) should be -81.

In [17]:
# Use this cell for your code
det_D = round(np.linalg.det(D))
print(det_D)
det_E = round(np.linalg.det(E))
print(det_E)

prod = round(np.linalg.det(D @ E))
print(prod)

-9
9
-81


-- *Use this markdown cell for your written answer* --

### 1.3 Slicing

Array slicing is a powerful way to extract data from an array. Let's practice array slicing with the following exercises!

1. Load the [California housing dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_california_housing.html). The data should be a matrix of shape <span style="background-color: #ccfff2">(20640, 8)</span>, that is 20640 rows and 8 columns. Use the <span style="background-color: #ccfff2">.shape</span> attribute of NumPy arrays to verify this. Here's a [description of the fields](https://scikit-learn.org/stable/datasets/real_world.html#california-housing-dataset).

In [18]:
# Use this cell for your code
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
h_df=housing.data
np.shape(h_df)

(20640, 8)

2. Select rows where the average number of bedrooms <span style="background-color: #ccfff2">(AveBedrms)</span> is higher than 2. The first few row indices should be <span style="background-color: #ccfff2">710,  1023,  1024, ...</span> (zero-indexed). Count these houses - how many rows are selected? *[hint: <span style="background-color: #ccfff2">np.where</span>]*

In [19]:
# Use this cell for your code

ave_bed = h_df[:, housing.feature_names.index("AveBedrms")]
indices = np.where(ave_bed >2)[0]
count = len(indices)
print(indices[:5])
count

[ 710 1023 1024 1030 1102]


235

3. Select the rows where the median house age (i.e. median in each block group) <span style="background-color: #ccfff2">(HouseAge)</span> is between 1 and 3 years (inclusive). There should be **124** of these.

In [20]:
# Use this cell for your code
house_age = h_df[:, housing.feature_names.index("HouseAge")]
indices = np.where((house_age >= 1) & (house_age <= 3))[0]
print("Total houses with age between 1 and 3:", len(indices))

Total houses with age between 1 and 3: 124


4. Find the mean of the block group population <span style="background-color: #ccfff2">(Population)</span> for homes whose median value is more than 25000 USD (the target variable). It should be around **1425.68**.

In [21]:
# Use this cell for your code
population = h_df[:, housing.feature_names.index("Population")]

target = housing.target

selected = population[target>0.25]

mean_pop = selected.mean()
print("Mean of the block group population:",mean_pop)

Mean of the block group population: 1425.6833737275813


## Exercise 2 | Working with text data

Next, let's look into some text data. We will be looking into Amazon reviews, and the necessary steps to transform a raw dataset into a format more suitable for prediction tasks.

1. Download the automotive 5-core dataset from [here](https://jmcauley.ucsd.edu/data/amazon_v2/categoryFilesSmall/Automotive_5.json.gz). Next, you can extract the data in <span style="background-color: #ccfff2">JSON</span> format. You can also download one of the bigger ones, if you are feeling ambitious. Open the JSON file. Access the <span style="background-color: #ccfff2">reviewText</span> field, which contains the unstructured review text written by the user.

For instance, the first review reads as follows: 

*'After I wrote the below review, the manufacturer contacted me and explained how to use this.  Instead of the (current) picture on Amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally. [...]'*

In [25]:
# Use this cell for your code
import json

reviews=[]
with open("Automotive_5.json","r", encoding="utf-8") as f:
    for line in f:
        review = json.loads(line)
        reviews.append(review.get("reviewText", ""))
    
#print(first_review)
print(reviews[0])

After I wrote the below review, the manufacturer contacted me and explained how to use this.  Instead of the (current) picture on Amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally.  Then the charge / sync cable goes through the hole and you can charge from the side.  When used in landscape position this stand does make more sense so I'm upgrading this review from 2 stars to 4 stars (which I rarely do, but in this case user error) with a star off for lack of documentation.  Manufacturer has said he will include directions and I would also encourage him to update the photos on Amazon with more examples of how this can be used.  The wood quality is indeed good and it's a sturdy little stand.  I'm trying to think of when I'd actually use this stand... perhaps for Facetime chats.  I also think it would be useful for someone in an open office environment if they want to keep their phone on their desk and see alerts / notifications easi

2. Next, let's follow some steps to normalize the text data.

When dealing with natural language, it is important to notice that while, for example, the words "Copper" and "copper" are represented by two different strings, they have the same meaning. When applying statistical methods on this data, it is useful to ensure that words with the same meaning are represented by the same string.

* <span style="background-color: #ccfff2">Downcasing</span>: Let's first downcase the contents of the <span style="background-color: #ccfff2">reviewText</span> field.

Now the first review should be:

*'after i wrote the below review, the manufacturer contacted me and explained how to use this.  instead of the (current) picture on amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally.'*

In [27]:
# Use this cell for your code
downcasted_reviews=[]

for text in reviews:
    downcasted_reviews.append(text.lower())
    
print(downcasted_reviews[0])

after i wrote the below review, the manufacturer contacted me and explained how to use this.  instead of the (current) picture on amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally.  then the charge / sync cable goes through the hole and you can charge from the side.  when used in landscape position this stand does make more sense so i'm upgrading this review from 2 stars to 4 stars (which i rarely do, but in this case user error) with a star off for lack of documentation.  manufacturer has said he will include directions and i would also encourage him to update the photos on amazon with more examples of how this can be used.  the wood quality is indeed good and it's a sturdy little stand.  i'm trying to think of when i'd actually use this stand... perhaps for facetime chats.  i also think it would be useful for someone in an open office environment if they want to keep their phone on their desk and see alerts / notifications easi

3. Let's continue with punctuation and stop word removal. Stop words are words like "and", "the", etc. They are usually very common words that have little to do with the actual content matter. There's plenty openly available lists of stop words for almost any (natural) language.

* <span style="background-color: #ccfff2">Punctuation and stop-word removal</span>: Let's now remove all punctuation, as well as the stop words. You can find a stop word list for English, e.g. [here](https://gist.github.com/xldrkp/4a3b1a33f10d37bedbe0068f2b4482e8#file-stopwords-en-txt).*(use the link to download a txt of english stopwords)* Save the stopwords in the file as "stopwords-en.txt".

First review at this point reads as: 

*'wrote review manufacturer contacted explained current picture amazon phone vertically stand phone horizontally'*

In [33]:
# Use this cell for your code
import string

with open("stopwords-en.txt", "r", encoding="utf-8") as f:
    stopwords = set(word.strip().lower() for word in f.readlines())
    
r_review = []

for text in downcasted_reviews:
    text = text.translate(str.maketrans("","",string.punctuation))
    
    words = [word for word in text.split() if word not in stopwords]
    r_review.append(" ".join(words))
    
print(r_review[0])

wrote review manufacturer contacted explained use instead current picture amazon phone placed vertically actually use stand phone placed horizontally charge sync cable goes hole charge used landscape position stand does make sense im upgrading review 2 stars 4 stars rarely case user error star lack documentation manufacturer said include directions encourage update photos amazon examples used wood quality good sturdy little stand im trying think id actually use stand facetime chats think useful open office environment want phone desk alerts notifications easily hope updated review helps original review throwing stand garbage crafted nice wood functional near tell does come instructions product description confusing english poorly translated chinese iphone sits stand just fine stand way plug plug stand blocks turn iphone plug just looks strange im sure hole stand supposed included plastic feet instructions affixed stand began wobble figure work thing post review far im concerned box exp

4. Let's continue with stemming. For example, while the words "swims" and "swim" are different strings, they both refer to swimming. [Stemming](https://en.wikipedia.org/wiki/Stemming) refers to the process of mapping words from their inflected form to their base form, for instance: swims -> swim.

* <span style="background-color: #ccfff2">Stemming</span>: Apply a stemmer on the paragraphs, so that inflected forms are mapped to the base form. For example, for Python the popular natural language toolkit [nltk](http://www.nltk.org/howto/stem.html) has an easy to use stemmer. In case you are using R, you can try the [Snowball stemmer](https://www.rdocumentation.org/packages/corpus/versions/0.10.2/topics/stem_snowball). You can find out how to install nltk from [here](https://www.nltk.org/install.html). It will take a while to run! So, grab a coffee and wait :D

Finally, after stemming: 

*'wrote review manufactur contact explain current pictur amazon phone vertic stand phone horizont'*

In [35]:
# Use this cell for your code
import nltk
from nltk.stem import PorterStemmer

stemmer = PorterStemmer()
stemmed_rev = []
for text in r_review:
    words = text.split()
    stemmed_words = [stemmer.stem(word) for word in words]
    stemmed_rev.append(" ".join(stemmed_words))

print(stemmed_rev[0])

wrote review manufactur contact explain use instead current pictur amazon phone place vertic actual use stand phone place horizont charg sync cabl goe hole charg use landscap posit stand doe make sens im upgrad review 2 star 4 star rare case user error star lack document manufactur said includ direct encourag updat photo amazon exampl use wood qualiti good sturdi littl stand im tri think id actual use stand facetim chat think use open offic environ want phone desk alert notif easili hope updat review help origin review throw stand garbag craft nice wood function near tell doe come instruct product descript confus english poorli translat chines iphon sit stand just fine stand way plug plug stand block turn iphon plug just look strang im sure hole stand suppos includ plastic feet instruct affix stand began wobbl figur work thing post review far im concern box experi terribl sorri recommend sampl provid exchang honest review


5. Finally, filter the data by selecting reviews where the field <span style="background-color: #ccfff2">overall</span> is 4 or 5, and store the review texts in a file named <span style="background-color: #ccfff2">pos.txt</span>. Similarly, select reviews with rating 1 or 2 and store them in a file named <span style="background-color: #ccfff2">neg.txt</span>. Ignore the reviews with overall rating 3. Each line in the two files should contain exactly one preprocessed review text without the rating.

In [40]:
# Use this cell for your code
ratings = []
with open("Automotive_5.json", "r", encoding="utf-8") as f:
    for line in f:
        review = json.loads(line)
        rating = review.get("overall", None)
        if text and rating is not None:
            ratings.append(rating)

with open("pos.txt", "w", encoding="utf-8") as pos_f, open("neg.txt", "w", encoding="utf-8") as neg_f:
    for review, rating in zip(stemmed_rev, ratings):
        if rating in [4, 5]:          # positive
            pos_f.write(review + "\n")
        elif rating in [1, 2]:        # negative
            neg_f.write(review + "\n")

In [43]:
with open("pos.txt", "r", encoding="utf-8") as f:
    pos_reviews = f.read().splitlines()

print("Total positive reviews:", len(pos_reviews))
print("First 3 positive reviews:\n", pos_reviews[:3])


Total positive reviews: 1476572
First 3 positive reviews:
 ['wrote review manufactur contact explain use instead current pictur amazon phone place vertic actual use stand phone place horizont charg sync cabl goe hole charg use landscap posit stand doe make sens im upgrad review 2 star 4 star rare case user error star lack document manufactur said includ direct encourag updat photo amazon exampl use wood qualiti good sturdi littl stand im tri think id actual use stand facetim chat think use open offic environ want phone desk alert notif easili hope updat review help origin review throw stand garbag craft nice wood function near tell doe come instruct product descript confus english poorli translat chines iphon sit stand just fine stand way plug plug stand block turn iphon plug just look strang im sure hole stand suppos includ plastic feet instruct affix stand began wobbl figur work thing post review far im concern box experi terribl sorri recommend sampl provid exchang honest review', '

In [44]:
with open("neg.txt", "r", encoding="utf-8") as f:
    neg_reviews = f.read().splitlines()

print("Total negative reviews:", len(neg_reviews))
print("First 3 negative reviews:\n", neg_reviews[:3])


Total negative reviews: 132298
First 3 negative reviews:
 ['suck bare pick definit car pretti dont wast money', 'write short blew 2 fuse car appar demand 12v', '2017 outback everytim tri use im blow fuse']


## Exercise 3 | SQL basics

Next, let's take a refresher on the basics of SQL. In this exercise, you will be working on the simplified Northwind 2000 SQLite database. You can download the database from Kaggle [here](https://courses.mooc.fi/api/v0/files/course/f92ffc32-2dd4-421d-87f3-c48800422cc5/files/VEKX2bxGCDGyojG902gmYZTXCnrAQw.zip).

To test your SQL queries and complete the exercise, you can download and install SQLite if you don't yet have it installed.

Please write SQL queries for the tasks on the simplified Northwind 2000 SQLite database.

1. List the first name, last name, and hire date of all employees hired after January 1st, 1994.

2. Count how many orders each customer has placed.

3. Find the names of all customers who have ordered the product "Chai".

4. Find all orders that have been placed but not yet shipped.

5. Find the customer who has placed the most orders.

**Remember to submit your solutions. You can return this Jupyter notebook (.ipynb) or .py, .R, etc depending on your programming preferences. Remember to also submit your SQL queries. No need to submit the text files for the programming exercises.**

In [45]:
import sqlite3
import pandas as pd

In [46]:
conn = sqlite3.connect("northwind2000-simplified.sqlite")

In [47]:
query = """
SELECT FirstName, LastName, HireDate
FROM Employees
WHERE HireDate > '1994-01-01';
"""
df = pd.read_sql_query(query, conn)
df.head()


FirstName   LastName    HireDate
0    Robert       King  1994-01-02
1     Laura   Callahan  1994-03-05
2      Anne  Dodsworth  1994-11-15

In [48]:
query = """
SELECT CustomerID, COUNT(OrderID) AS OrderCount
FROM Orders
GROUP BY CustomerID
ORDER BY OrderCount DESC;
"""
df = pd.read_sql_query(query, conn)
df.head()


CustomerID  OrderCount
0      SAVEA          31
1      ERNSH          30
2      QUICK          28
3      HUNGO          19
4      FOLKO          19

In [49]:
query = """
SELECT DISTINCT c.CompanyName
FROM Customers c
JOIN Orders o ON c.CustomerID = o.CustomerID
JOIN OrderDetails od ON o.OrderID = od.OrderID
JOIN Products p ON od.ProductID = p.ProductID
WHERE p.ProductName = 'Chai';
"""
df = pd.read_sql_query(query, conn)
df.head()


CompanyName
0                  QUICK-Stop
1  Rattlesnake Canyon Grocery
2    Lonesome Pine Restaurant
3           Die Wandernde Kuh
4   Pericles Comidas clásicas

In [50]:
query = """
SELECT OrderID, CustomerID, OrderDate
FROM Orders
WHERE ShippedDate IS NULL;
"""
df = pd.read_sql_query(query, conn)
df.head()


OrderID CustomerID                OrderDate
0    11008      ERNSH  1998-04-08 00:00:00.000
1    11019      RANCH  1998-04-13 00:00:00.000
2    11039      LINOD  1998-04-21 00:00:00.000
3    11040      GREAL  1998-04-22 00:00:00.000
4    11045      BOTTM  1998-04-23 00:00:00.000

In [51]:
query = """
SELECT c.CustomerID, c.CompanyName, COUNT(o.OrderID) AS OrderCount
FROM Customers c
JOIN Orders o ON c.CustomerID = o.CustomerID
GROUP BY c.CustomerID, c.CompanyName
ORDER BY OrderCount DESC
LIMIT 1;
"""
df = pd.read_sql_query(query, conn)
df


CustomerID         CompanyName  OrderCount
0      SAVEA  Save-a-lot Markets          31